File: lib/s/prims.ts
import {
  vFalse,
  vInt,
  vTrue,
  valEq,
  withVal,
  type Val,
} from "@/lib/libamp/values"

/**
 * Primitive registry. Corresponds to `O[[-]] : Prim -> Val* -> Val`.
 * Primitives are total on valid argument shapes and throw `PrimError` otherwise;
 * the CEK machine catches these to produce a `Stuck` trace step.
 */

export class PrimError extends Error {
  constructor(message: string) {
    super(message)
  }
}

export type Prim = (args: Val[]) => Val

function expectInt(v: Val, op: string, idx: number): number {
  return withVal(v, {
    int: ({ n }) => n,
    ctor: ({ tag }) => {
      throw new PrimError(
        `primitive ${op}: argument ${idx} expected int, got ${tag}(...)`
      )
    },
  })
}

function expectArity(op: string, args: Val[], n: number): void {
  if (args.length !== n) {
    throw new PrimError(
      `primitive ${op}: expected ${n} argument(s), got ${args.length}`
    )
  }
}

export const PRIMS: Record<string, Prim> = {
  sub: (args) => {
    expectArity("sub", args, 2)
    return vInt(expectInt(args[0], "sub", 0) - expectInt(args[1], "sub", 1))
  },
  add: (args) => {
    expectArity("add", args, 2)
    return vInt(expectInt(args[0], "add", 0) + expectInt(args[1], "add", 1))
  },
  mul: (args) => {
    expectArity("mul", args, 2)
    return vInt(expectInt(args[0], "mul", 0) * expectInt(args[1], "mul", 1))
  },
  iszero: (args) => {
    expectArity("iszero", args, 1)
    return expectInt(args[0], "iszero", 0) === 0 ? vTrue : vFalse
  },
  eq: (args) => {
    expectArity("eq", args, 2)
    return valEq(args[0], args[1]) ? vTrue : vFalse
  },
  lt: (args) => {
    expectArity("lt", args, 2)
    return expectInt(args[0], "lt", 0) < expectInt(args[1], "lt", 1)
      ? vTrue
      : vFalse
  },
  not: (args) => {
    expectArity("not", args, 1)
    return withVal(args[0], {
      int: () => {
        throw new PrimError(`primitive not: expected boolean, got non-boolean`)
      },
      ctor: ({ tag }) => {
        if (tag !== "True" && tag !== "False") {
          throw new PrimError(
            `primitive not: expected boolean, got non-boolean`
          )
        }
        return tag === "True" ? vFalse : vTrue
      },
    })
  },
}

export function isPrim(name: string): boolean {
  return Object.prototype.hasOwnProperty.call(PRIMS, name)
}

File: lib/s/prims.ml
open Values

type prim_result = (value, string) result

let ( let* ) = Result.bind

let expectArity op args n =
  if Array.length args = n then Ok ()
  else
    Error
      ("primitive " ^ op ^ ": expected " ^ string_of_int n
     ^ " argument(s), got "
      ^ string_of_int (Array.length args))

let expectInt op idx = function
  | Int { n } -> Ok n
  | Ctor { tag; _ } ->
      Error
        ("primitive " ^ op ^ ": argument " ^ string_of_int idx
       ^ " expected int, got " ^ tag ^ "(...)")

let expectBoolean op = function
  | Ctor { tag = ("True" | "False") as tag; _ } -> Ok tag
  | _ -> Error ("primitive " ^ op ^ ": expected boolean, got non-boolean")

let boolValue b = if b then vTrue else vFalse

let evalPrim op args =
  match (op, args) with
  | "sub", [ v1; v2 ] ->
      let* n1 = expectInt "sub" 0 v1 in
      let* n2 = expectInt "sub" 1 v2 in
      Ok (vInt (n1 - n2))
  | "add", [ v1; v2 ] ->
      let* n1 = expectInt "add" 0 v1 in
      let* n2 = expectInt "add" 1 v2 in
      Ok (vInt (n1 + n2))
  | "mul", [ v1; v2 ] ->
      let* n1 = expectInt "mul" 0 v1 in
      let* n2 = expectInt "mul" 1 v2 in
      Ok (vInt (n1 * n2))
  | "iszero", [ v ] ->
      let* n = expectInt "iszero" 0 v in
      Ok (boolValue (n = 0))
  | "eq", [ v1; v2 ] ->
      let* n1 = expectInt "eq" 0 v1 in
      let* n2 = expectInt "eq" 1 v2 in
      Ok (boolValue (n1 = n2))
  | "lt", [ v1; v2 ] ->
      let* n1 = expectInt "lt" 0 v1 in
      let* n2 = expectInt "lt" 1 v2 in
      Ok (boolValue (n1 < n2))
  | "not", [ v ] ->
      let* tag = expectBoolean "not" v in
      Ok (if tag = "True" then vFalse else vTrue)
  | _ -> Error ("unknown primitive or wrong arity: " ^ op)
